### PR TITLE
Fix ReturnInVoidContext cop when return is not inside a method.

### PR DESCRIPTION
**Pull request by frodsan**
Fixes https://github.com/bbatsov/rubocop/issues/4736

### CODE REVIEW

File: lib/rubocop/cop/lint/return_in_void_context.rb [hunk 1]
```diff
@@ -45,7 +45,7 @@ def on_return(return_node)
 
         def method_name(return_node)
```
Comment (Drenmi) [inline lib/rubocop/cop/lint/return_in_void_context.rb]: I am inclined to split this into two methods, one that finds the context, and one that extracts the name from a `def` node.

Right now these two are mixed, and causes this method to have a "special meaning `nil`" mixed return type.

I would rather see something like:

```ruby
context_node = non_void_context(return_node)

return unless context_node && ...

add_offense(return_node, :keyword, format(message, method_name(context_node))
```

File: CHANGELOG.md [hunk 2]
```diff
@@ -2,6 +2,10 @@
 
 ## master (unreleased)
 
+### Bug fixes
+
+* [#4737](https://github.com/bbatsov/rubocop/pull/4737): Fix ReturnInVoidContext cop when `return` is not inside a method. ([@frodsan][])
```
Comment (Drenmi) [inline CHANGELOG.md]: Technically "not inside a context", or "in top scope". 🙂

File: spec/rubocop/cop/lint/return_in_void_context_spec.rb [hunk 3]
```diff
@@ -80,4 +80,12 @@ def self.initialize
       RUBY
     end
   end
+
+  context 'with no method' do
```
Comment (Drenmi) [inline spec/rubocop/cop/lint/return_in_void_context_spec.rb]: Here, too: "when `return` is in top scope".

File: CHANGELOG.md (hunk 2 above)
Comment (bbatsov) [reply CHANGELOG.md]: You can use `return` in any scope, so the description seems reasonable to me, unless I missing something @Drenmi.

File: spec/rubocop/cop/lint/return_in_void_context_spec.rb (hunk 3 above)
Comment (bbatsov) [reply spec/rubocop/cop/lint/return_in_void_context_spec.rb]: Guess we can also have some tests with procs, lambda and a return in say class definition. Guess we didn't think of all of them in the past.

File: CHANGELOG.md (hunk 2 above)
Comment (bbatsov) [reply CHANGELOG.md]: Actually I'm wrong - I tested this and you can use it only in method/lambda/proc or in the top scope. My bad.

File: spec/rubocop/cop/lint/return_in_void_context_spec.rb (hunk 3 above)
Comment (Drenmi) [reply spec/rubocop/cop/lint/return_in_void_context_spec.rb]: Good point!

**Comment by frodsan** [comment]
@Drenmi updated

File: lib/rubocop/cop/lint/return_in_void_context.rb [hunk 4]
```diff
@@ -33,30 +33,39 @@ module Lint
...
-            !AST::Node::COMPARISON_OPERATORS.include?(method_name)
+        def method_name(context_node)
+          context_node.children.first
         end
 
         def useless_return_method?(method_name)
           method_name == :initialize || method_setter?(method_name)
         end
+
+        def method_setter?(method_name)
```
Comment (Drenmi) [inline lib/rubocop/cop/lint/return_in_void_context.rb]: `#method_setter?` -> `#setter_method?` 🙂

File: lib/rubocop/cop/lint/return_in_void_context.rb [hunk 5]
```diff
@@ -33,30 +33,39 @@ module Lint
...
         end
 
-        def method_setter?(method_name)
-          method_name.to_s.end_with?('=') &&
-            !AST::Node::COMPARISON_OPERATORS.include?(method_name)
+        def method_name(context_node)
+          context_node.children.first
         end
 
         def useless_return_method?(method_name)
```
Comment (Drenmi) [inline lib/rubocop/cop/lint/return_in_void_context.rb]: Should this be `#void_context_method?`

**Comment by frodsan** [comment]
@Drenmi updated

**Comment by bbatsov** [comment]
Thanks!

**Pull request merged**